mala/datahandling/multi_lazy_load_data_loader.py

```diff
@@ -0,0 +1,180 @@
+        # dataset in list
+        dset = self.datasets[0]
+        dset.allocate_shared_mem()
+        self.load_future = self.pool.submit(self.load_snapshot_to_shm,
+                                            dset.snapshot,
+                                            dset.descriptor_calculator,
+                                            dset.target_calculator,
+                                            dset.input_shm_name,
+                                            dset.output_shm_name)
+
+    def __len__(self):
+        """
+        Access number of datasets/snapshots contained within this loader.
+
+        Returns
+        -------
+        length : int
+            Number of datasets/snapshots contained within this loader.
+        """
+        return len(self.loaders)
+
+    def __iter__(self):
+        """
+        Generate an iterator.
+
+        Returns
+        -------
+        iterator: MultiLazyLoadDataLoader
+            An iterator over the individual datasets/snapshots in this
+            object.
+        """
+        self.count = 0
+        return self
+
+    def __next__(self):
+        """
+        Access the next data loader.
+
+        Returns
+        -------
+        iterator: DataLoader
+            The next data loader.
+        """
+        self.count += 1
+        if self.count > len(self.loaders):
+            raise StopIteration
+        else:
+            # Wait on last prefetch
+            if self.count - 1 >= 0:
+                if not self.datasets[self.count - 1].loaded:
+                    self.load_future.result()
+                    self.datasets[self.count - 1].loaded = True
+
+            # Delete last
+            if self.count - 2 >= 0:
+                self.datasets[self.count - 2].delete_data()
+
+            # Prefetch next file (looping around epoch boundary)
+            dset = self.datasets[self.count % len(self.loaders)]
+            if not dset.loaded:
+              dset.allocate_shared_mem()
+              self.load_future = self.pool.submit(self.load_snapshot_to_shm,
+                                                  dset.snapshot,
+                                                  dset.descriptor_calculator,
+                                                  dset.target_calculator,
+                                                  dset.input_shm_name,
+                                                  dset.output_shm_name)
+
+            # Return current
+            return self.loaders[self.count - 1]
+
+    # TODO: Without this function, I get 2 times the number of snapshots
+    # memory leaks after shutdown. With it, I get 1 times the number of
+    # snapshots. So using this function seems to be correct; just simply not
+    # enough? I am not sure where the memory leak is coming from.
+    def cleanup(self):
+        """Deallocate arrays still left in memory."""
+        for dset in self.datasets:
+            dset.delete_data()
+
```

Comment:
In trying to deal with the leaked `shared_memory` warnings at job termination, I added:
```
self.pool.shutdown()
```
to stop the process pool before shutdown in this `cleanup` method. This didn't resolve the warnings, but is something that should be added anyway.

I wonder if we are running into the Python bug mentioned in this StackOverflow post?: https://stackoverflow.com/questions/62748654/python-3-8-shared-memory-resource-tracker-producing-unexpected-warnings-at-appli

Reply:
I just tested this and it removed two of the three warnings (the weakref ones) for me, thanks a lot! :) 

As for the Stack Overflow post: I found that too and think you might be right, that could be related...

Reply:
Fixed as per @romerojosh suggestions. The PR is ready to be merged now.